### DISCUSSION TITLE
Use of generics for typing collections and documents

### DISCUSSION THREAD

**Discussion by nthonymiller** [firebase/firebase-js-sdk#4404] [Q&A]
You have the API:
```typescript
 collection(restaurantRef, 'ratings')
```
Can this be strongly typed by using typescript generics?

**Comment by samtstern** [comment]
@nthonymiller yes both `DocumentReference` and `CollectionReference` are generic and can be typed by providing a `FirestoreDataConverter`

So for example let's say I have a `City` class:

```ts
const cityConverter = {
    toFirestore: (city: City) => {
        return {
            name: city.name,
            state: city.state,
            country: city.country
            };
    },
    fromFirestore: (snapshot, options) => {
        const data = snapshot.data(options);
        return new City(data.name, data.state, data.country);
    }
};

const ref = doc(collection(db, "cities"), "LA").withConverter(cityConverter);
```

In this case `ref` would be of type `DocumentReference<City>` so ` get()` or `set()` operation will have the appropriate types.

**Reply by nthonymiller** [reply]
@samtstern are you saying that the usage of the converter would infer the type?

What about if you don't use a converter? What if the return is not a class instance but an interface?

```typescript

interface City {
   name: string;
   state: string;
   country: string;
}

const ref = doc(collection<City>(db, "cities"), "LA");

```
The doc type could be inferred from the Collection type.

**Reply by samtstern** [reply]
It can be an `interface` rather than a `class` but you should use a Converter.

The simplest converter could just be a cast:
```ts
const cityConverter = {
    toFirestore: (city: City) => {
        return city;
    },
    fromFirestore: (snapshot, options) => {
        const data = snapshot.data(options);
        return data as City;
    }
}
```

**Reply by samtstern** [reply]
Also the doc type will be inferred from the collection type in your example, but you need to use a converter somewhere in the change to get the first typed reference and then build off thrat.

**Reply by nthonymiller** [reply]
OK I understand. 

I would like to make a suggestion here: I think the usage of a converter in order to get to a type is not optimal. I have found not many dev's use converters at all.

**Reply by vis97c** [reply]
collection function should be generic
